### PR TITLE
add a formula for "redpen"

### PR DESCRIPTION
**Pull request by kozo2**
RedPen is a proofreading tool to help writers or programmers who write
technical documents or manuals that need to adhere to a writing
standard. RedPen automates the verifications of input documents written
in natural languages.

### CODE REVIEW

**Comment by griffin-stewie** [comment]
+1

**Comment by mikemcquaid** [comment]
`brew test` and `brew audit --strict` need fixed before this can be included: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/19349/testReport/

**Comment by kozo2** [comment]
@mikemcquaid 
Thank you for your comment. I fixed `brew test` and `brew audit --strict` problem.

**Comment by mikemcquaid** [comment]
> Thank you for your comment. It seems brew test needs JAVA_HOME value, like JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk1.8.0_25.jdk/Contents/Home brew test redpen.
> How can I set this value in redpen.rb?

`bin.env_script_all_files(libexec/"bin", :JAVA_HOME => "`/usr/libexec/java_home`")` to write the wrapper script instead of installing a symlink.

**Comment by kozo2** [comment]
Thank you @mikemcquaid , I think now the `JAVA_HOME` problem is fixed.

**Comment by takahi-i** [comment]
+1

**Comment by mikemcquaid** [comment]
@kozo2 Still waiting on a better test, thanks.

File: Library/Formula/redpen.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  depends_on :java => "1.8"
+
+  def install
+    # Don't need Windows files.
+    rm_f Dir["bin/*.bat"]
+    libexec.install %w[conf lib sample-doc]
+
+    prefix.install "bin"
+    bin.env_script_all_files(libexec/"bin", :JAVA_HOME => "`/usr/libexec/java_home`")
```
Comment (mikemcquaid) [inline Library/Formula/redpen.rb]: Remove the double-quotes around `/usr/libexec/java_home`

Comment (mikemcquaid) [inline Library/Formula/redpen.rb]: You probably also want to `.chomp` that output:
```ruby
java_home = `/usr/libexec/java_home`.chomp
bin.env_script_all_files(libexec/"bin", :JAVA_HOME => java_home)
```

**Comment by kozo2** [comment]
Thanks @mikemcquaid 
I removed double-quotes and added `.chomp`.

**Comment by mikemcquaid** [comment]
@kozo2 I'd still like to see a better `test do` block.

**Comment by takahi-i** [comment]
@kozo2  How about adding the following line in the **test** block?

    system "#{bin}/redpen", "-c", "conf/redpen-conf-en.xml", "-l", "10", "sample-doc/en/sampledoc-en.txt"

The above line executes redpen with the bundled sample configuration and text.

**Comment by mikemcquaid** [comment]
Yup, that test would be great!

**Comment by kozo2** [comment]
Thanks @takahi-i , I used your advise written in https://gist.github.com/takahi-i/fece21865e4a189e22f6

@mikemcquaid I added test for first ValidationError type in sampledoc-en.txt

File: Library/Formula/redpen.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    rm_f Dir["bin/*.bat"]
+    libexec.install %w[conf lib sample-doc]
+
+    prefix.install "bin"
+    java_home = `/usr/libexec/java_home`.chomp
+    bin.env_script_all_files(libexec/"bin", :JAVA_HOME => java_home)
+  end
+
+  test do
+    system "#{bin}/redpen", "-version"
```
Comment (mikemcquaid) [inline Library/Formula/redpen.rb]: Can get rid of this.

File: Library/Formula/redpen.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+    file_name = "sampledoc-en.txt"
+    path = "#{libexec}/sample-doc/en/#{file_name}"
```
Comment (mikemcquaid) [inline Library/Formula/redpen.rb]: Can just inline this variable.

File: Library/Formula/redpen.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    output = `#{bin}/redpen -c #{libexec}/conf/redpen-conf-en.xml #{path}`
```
Comment (mikemcquaid) [inline Library/Formula/redpen.rb]: `assert_matches /#{file_name}:1: ValidationError[SymbolWithSpace]/, system_output("#{bin}/redpen ..."` can replace this line and the two below.

**Comment by kozo2** [comment]
Thanks @mikemcquaid 
I reflected your comment in https://github.com/Homebrew/homebrew/commit/e94cedbf334a4fcce53357f4e345577dd4978253

**Comment by diimdeep** [comment]
```
$ brew cask install java
$ brew install redpen
$ redpen
12:07:22.186 [main] INFO  cc.redpen.ConfigurationLoader - Loading config from specified config file: "/usr/local/Library/Formula"
Exception in thread "main" cc.redpen.RedPenException:  (No such file or directory)
	at cc.redpen.ConfigurationLoader.load(ConfigurationLoader.java:102)
	at cc.redpen.RedPen.<init>(RedPen.java:54)
	at cc.redpen.Main.run(Main.java:158)
	at cc.redpen.Main.main(Main.java:64)
Caused by: java.io.FileNotFoundException:  (No such file or directory)
	at java.io.FileInputStream.open(Native Method)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at cc.redpen.ConfigurationLoader.load(ConfigurationLoader.java:99)
	... 3 more
```

**Comment by takahi-i** [comment]
Hi Dmitry,

Thank you very much for trying RedPen.

You need to run the redpen command with configuration file and input file.
You can run the command with the bundled resources as follows.

redpen -c /usr/local/Cellar/redpen/1.0.1/libexec/conf/redpen-conf-en.xml
/usr/local/Cellar/redpen/1.0.1/libexec/sample-doc/en/sampledoc-en.txt

Best,
Takahiko

2015-01-28 18:10 GMT+09:00 Dmitry Wolf <notifications@github.com>:

> $ brew cask install java
> $ brew install redpen
> $ redpen
> 12:07:22.186 [main] INFO  cc.redpen.ConfigurationLoader - Loading config from specified config file: "/usr/local/Library/Formula"
> Exception in thread "main" cc.redpen.RedPenException:  (No such file or directory)
>     at cc.redpen.ConfigurationLoader.load(ConfigurationLoader.java:102)
>     at cc.redpen.RedPen.<init>(RedPen.java:54)
>     at cc.redpen.Main.run(Main.java:158)
>     at cc.redpen.Main.main(Main.java:64)
> Caused by: java.io.FileNotFoundException:  (No such file or directory)
>     at java.io.FileInputStream.open(Native Method)
>     at java.io.FileInputStream.<init>(FileInputStream.java:138)
>     at cc.redpen.ConfigurationLoader.load(ConfigurationLoader.java:99)
>     ... 3 more
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Homebrew/homebrew/pull/35147#issuecomment-71802052>.
>



-- 
Takahiko

**Comment by diimdeep** [comment]
@takahi-i  thanks, better show human readable error instead of exception stack

**Comment by takahi-i** [comment]
Thank you for the comment @diimdeep. I added a new issue to make error
human readable.

https://github.com/recruit-tech/redpen/issues/383

2015-01-28 19:23 GMT+09:00 Dmitry Wolf <notifications@github.com>:

> @takahi-i <https://github.com/takahi-i> thanks, better show human
> readable error instead of exception stack
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Homebrew/homebrew/pull/35147#issuecomment-71811396>.
>



-- 
Takahiko

**Pull request closed without merging**